eventcrypto.go

```diff
@@ -0,0 +1,161 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ed25519"
+)
+
+// AddContentHashesToEvent sets the "hashes" key of the event with a SHA-256 hash of the unredacted event content.
+// This hash is used to detect whether the unredacted content of the event is valid.
+func AddContentHashesToEvent(eventJSON []byte) ([]byte, error) {
+	var event map[string]rawJSON
+
+	if err := json.Unmarshal(eventJSON, &event); err != nil {
+		return nil, err
+	}
+
+	unsignedJSON := event["unsigned"]
+
+	delete(event, "signatures")
+	delete(event, "unsigned")
```

Comment:
Setting the precedent for using constants here would be a good idea IMO to prevent fat-fingering this. I appreciate we have good test coverage such that this would be immediately obvious, but I'm mainly thinking of the future when things inevitably get added. People will do what the surrounding code does, and if the surrounding code uses constants, so will they. This decreases the chance of them extracting the right key but then deleting a different key. The downside is that we're adding in a layer of indirection which may be unclear.

WDYT?